<a href="https://colab.research.google.com/github/gpasky/MAT-421/blob/main/ModD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Gabriel Pascual

**1.2.1 Linear Spaces**

A linear subspace is the result of a linear combination that multiplies each vector from a subset by a constant and sums up the end results.

Definition of Linear subspace:
$V$ is a subset $U ⊆ V$

$\forall$ $u_1,u_2 \in U,$ $and$ $αu_1 \in U$

A linear subspace always contains 0

1.2.2

Span: The set of all linear combinations of the $w_j$'s

$span(w_1,...,w_m)$ = ${∑_{j=1}^{m}} a_jw-j: a_1,...a_m\in\mathbb{R}$

*   A list of vectors could span a linear subspace

A span is a linear subspace:

Let $W = span(w_1,...,w_m)$ = **linear subspace**

The conditions $u_1,u_2 \in U,$ $and$ $αu_1 \in U$ are satisfied for a linear subspace since

$au_1+u_2 = ∑_{j=1}^{m}(aβ_1,j+β_2,j)w_j$ given i = 1,2.

This represents a linear combination from combining both conditions






1.2.4

Column space: Given a matrix of $n$ x $m$ with columns $a_1...,a_m\in \mathbb{R^n}$, the linear combination of these vectors is $col(A) = span(a_1,...,a_m)$.

1.2.1.2

Linear Independence and Dimension:

*   Linear independence occurs if none of the vectors $u_1...,u_m$ can be a linear combination of the others

$∀i, u_i ∉ span({u_j: j\not_= i})$

*   Linear dependence occurs if $\sum_{j=1}^{m}a_ju_j=0$ given that $∃a_j$ not all zero, whereas independence occurs if the condition implies $a_j=0, ∀j$.

Ex: v = [1,2,0], w = [0,0,3], u = [2,0,0].
Then w is linearly independent from v and u since its third element is greater than 0, v and u are linearly independent since v contains a value greater than zero on its second element.

*   Dimension Theorem: When given U to be a linear subspace of V, the basis contains the same number of elements (dimension or dim(U))

**1.2.2 Orthogonality**

1.2.2.1

Orthonormal Bases: Vectors ${u_1,...,u_m}$ are orthonormal if $u_i$'s are pairwise orthogonal(perpendicular) and each has a norm equal to 1:

*   $<u_i,u_j> = u_i * u_j=0$
*   $\|u_i\|=1$

An example can be made using vectors v = [0,3,8] and w = [-10,-8,3].

In [2]:
import numpy as np
from numpy.linalg import norm
from numpy import arccos, dot

v = np.array([[0,3,8]])
w = np.array([[-10,-8,3]])
theta = \
    arccos(dot(v,w.T)/(norm(v)*norm(w)))
print(theta)

[[1.57079633]]


We get a result of $1.57079633 ≈ π / 2$ radians which means that the vectors are orthogonal.

1.2.2.2

Best Approximation Theorem:
Given a linear subspace $S ⊆ V$ and vector v not in that subspace, we find v* in S closest to v$.

Ex: $S=span(u_1)$ where $\| u_1\| = 1$. Solving for v* we see that v-v* is orthogonal to $u_1$.

Let $v^*=a^*u_1$

*   $0=<u_1,v-v^*>$
*   $0=<u_1,v>$ - $a^*$
*   $a^*=<u_1,v>$

Thus,  $v^*=<u_1,v>u_1$

**1.2.4 Eigenvalues and Eigenvectors**

$Ax=λx$ where $λ$ is an eigenvalue of matrix A if it holds true with the matrix A. The x vector is an eigenvector associated to $λ$. Eigenvectors transform Ax while the eigenvalues are the scale factors for this transformation.

An example can be made to find both eigenvalues and eigenvectors of the matrix $A=\begin{bmatrix}0&2\\2&3\end{bmatrix}$


In [7]:
import numpy as np

A = np.array([[0,3],[3,4]])

eigenvalue,eigenvector = np.linalg.eig(A)

print("Eigenvalue:",eigenvalue,"\nEigenvector:",eigenvector)

Eigenvalue: [-1.60555128  5.60555128] 
Eigenvector: [[-0.8816746  -0.47185793]
 [ 0.47185793 -0.8816746 ]]


The Power Method can also be used to find the largest eigenvalue with its eigenvector. This iterative method takes the n x n matrix A with $λ_n$ eigenvalues and $v_1,...v_n$ eigenvectors with the contraint $|λ_1| > |λ_2|$.

$Ax_{k-1}=λ_1[v_1+\frac{c_2}{c_1}\frac{\lambda_{2}^{k}}{\lambda_{1}^{k}}v_2+...+v_1+\frac{c_n}{c_n}\frac{\lambda_{n}^{k}}{\lambda_{n}^{k}}v_n$



*   For a large k, we eventually end up with the largest eigenvalue

*   By factoring out the largest element in the vector of each iteration, we normalize to get the vector equal to 1

An example can be made to show this iterative process:

In [8]:
import numpy as np

def normalize(x):
  fac = abs(x).max()
  x_n = x / x.max()
  return fac, x_n

x = np.array([1,1])
a = np.array([[0,3],[3,4]])

for i in range(8):
  x = np.dot(a,x)
  lambda_1, x = normalize(x)

print("Eigenvalue:",lambda_1,"\nEigenvector:",x)

Eigenvalue: 5.60536649529986 
Eigenvector: [0.5352014 1.       ]
